compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionUtils.cpp

```diff
@@ -56,6 +63,118 @@ static bool canInsOperandTieWithOutsOperand(OpOperand *insOperand) {
   return llvm::any_of(linalgOp.getOutputOperands(), canTieWithOutsOperand);
 }
 
+/// Checks if a linalg op is a simple reduction of the innermost dimensions
+/// with identity map for the input.
+static bool isReductionOnInnermostDims(linalg::LinalgOp linalgOp) {
+  SmallVector<Operation *, 4> combinerOps;
+
+  auto numInputs = linalgOp.getNumInputs();
+  if (numInputs != 1 && numInputs != 2) return false;
+
+  if (linalgOp.getNumOutputs() != 1) return false;
+
+  // Check if the result dims are d0, d1, ..., which means the reduction is done
+  // in the innermost dimensions.
+  auto output = linalgOp.getOutputOperand(0);
+  auto outputIndexingMap = linalgOp.getTiedIndexingMap(output);
+  for (const auto &en : llvm::enumerate(outputIndexingMap.getResults())) {
+    auto expr = en.value();
+    if (auto dim = expr.dyn_cast<AffineDimExpr>()) {
+      if (dim.getPosition() != en.index()) return false;
+    } else {
+      return false;
+    }
+  }
+
+  if (!matchReduction(linalgOp.getRegionOutputArgs(), 0, combinerOps) ||
+      combinerOps.size() != 1) {
+    return false;
+  }
+
+  return true;
+}
+
+/// Check if the op is elementwise and the output indexing map is identity.
+static bool isSimpleElementwise(linalg::LinalgOp op) {
+  if (op.getNumLoops() != op.getNumParallelLoops()) return false;
+
+  if (!allIndexingsAreProjectedPermutation(op)) return false;
+
+  for (OpOperand *opOperand : op.getOutputOperands()) {
+    if (!op.getTiedIndexingMap(opOperand).isIdentity()) return false;
+  }
+  return linalg::hasOnlyScalarElementwiseOp(op->getRegion(0));
+}
+
+/// Check if the use of operand is a reduction -> broadcast -> elementwise.
+/// An example:
+///   s = sum(a: <16x32xf32>) -> <16xf32>
+///   d = div(a: <16x32xf32>, broadcast(s to <16x32xf32>)) -> <16x32xf32>
+static bool isReductionBroadcastElementwise(OpOperand *operand) {
```

Comment:
It is great to see that relatively simple conditions / filters allow creating more powerful dispatch regions!

I am not clear about the structure of the code here though, these are somewhat adhoc constraints that are conjunctions of things hardcoded deep in C++. I would love to see a common effort towards more composable constraints that we can reuse and reassemble better, esp. as we decouple more and more the control from hardcoded C++.

I'd encourage you to think about refactoring to allow easily plugging and composing with the PR I sent out here: https://github.com/iree-org/iree/pull/9756 (last commit).

Another way to phrase my request is: "what are the atomic units of useful matching and filtering" that we can reuse; isReductionBroadcastElementwise looks to me like a very non-atomic that we should build out of atoms that we want to properly define to make them reusable.

Reply:
Thanks @nicolasvasilache! I agree. When I worked on this to recover the original tensor op's info, I felt like I am trying recovering the upper IR level representation (tosa or mhlo) at the linalg generic op level reversely.

This led me to thinking of the new compilation pipeline, which needs a common high level tensor ops, such as tosa, mhlo, or purely linalg name ops before linalg. This could have made the pattern matching code a lot easy and clean. This would need a great amount of discussion, so I will initiate the discussion later. 

I am further working on the reduction code to support dispatch fusions including softmax and layer normalization, and I will put the refactoring effort while working on them by referencing your PR. Thanks for the pointer!